include/exadg/operators/navier_stokes_calculators.cpp

```diff
@@ -84,10 +144,232 @@ ShearRateCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> cons
+  {
+    integrator_vector.reinit(cell);
+    integrator_vector.gather_evaluate(src, dealii::EvaluationFlags::values);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_scalar.n_q_points; q++)
+    {
+      scalar magnitude = integrator_vector.get_value(q).norm();
+      integrator_scalar.submit_value(magnitude, q);
+    }
+    integrator_scalar.integrate_scatter(dealii::EvaluationFlags::values, dst);
+  }
+}
+
+template<int dim, typename Number>
+QCriterionCalculator<dim, Number>::QCriterionCalculator()
+  : matrix_free(nullptr),
+    dof_index_u(0),
+    dof_index_u_scalar(0),
+    quad_index(0),
+    compressible_flow(false)
+{
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::initialize(
+  dealii::MatrixFree<dim, Number> const & matrix_free_in,
+  unsigned int const                      dof_index_u_in,
+  unsigned int const                      dof_index_u_scalar_in,
+  unsigned int const                      quad_index_in,
+  bool const                              compressible_flow_in)
+{
+  matrix_free        = &matrix_free_in;
+  dof_index_u        = dof_index_u_in;
+  dof_index_u_scalar = dof_index_u_scalar_in;
+  quad_index         = quad_index_in;
+  compressible_flow  = compressible_flow_in;
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::compute(VectorType & dst, VectorType const & src) const
+{
+  dst = 0;
+
+  matrix_free->cell_loop(&This::cell_loop, this, dst, src, compressible_flow);
+}
+
+template<int dim, typename Number>
+void
+QCriterionCalculator<dim, Number>::cell_loop(dealii::MatrixFree<dim, Number> const & matrix_free,
+                                             VectorType &                            dst,
+                                             VectorType const &                      src,
+                                             Range const & cell_range) const
+{
+  CellIntegratorVector integrator_vector(matrix_free, dof_index_u, quad_index);
+  CellIntegratorScalar integrator_scalar(matrix_free, dof_index_u_scalar, quad_index);
+
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    integrator_vector.reinit(cell);
+    integrator_vector.gather_evaluate(src, dealii::EvaluationFlags::gradients);
+
+    integrator_scalar.reinit(cell);
+
+    for(unsigned int q = 0; q < integrator_scalar.n_q_points; q++)
+    {
+      tensor grad_u = integrator_vector.get_gradient(q);
+      tensor Om, S;
+      for(unsigned int i = 0; i < dim; i++)
+      {
+        for(unsigned int j = 0; j < dim; j++)
+        {
+          Om[i][j] = 0.5 * (grad_u[i][j] - grad_u[j][i]);
+          S[i][j]  = 0.5 * (grad_u[i][j] + grad_u[j][i]);
+        }
+      }
+
+      // Q criterion for compressible flow is based on the deviatoric
+      // part of S (Kolar, V.: Compressibility Effect in Vortex Identification,
+      // AIAA, 2009. https://doi.org/10.2514/1.40131)
+      if(compressible_flow)
```

Comment:
here is the if at the quadrature point level:
1) leave it that way
2) duplicate cell loop inside `cell_loop` and have an if at the front
3) ?
What should we go for?